In [1]:
import numpy as np
import pandas as pd
from dask import dataframe as dd
import string
import re
import nltk
from nltk.stem.snowball import SnowballStemmer
from nltk.corpus import stopwords
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English

nlp = English()
tokenizer = nlp.tokenizer
stemmer = SnowballStemmer("english")

pd.set_option('display.max_columns', 100)

# Government Spend Analysis

## Step 1: Examine column headers to identify columns of interest

In [2]:
test = pd.read_csv('FY2016_All_Contracts_Full_20210608_1.csv', nrows = 5)

test

contract_transaction_unique_key             contract_award_unique_key  \
0  8600_8600_0001_0_DU205NC16D02_0  CONT_AWD_0001_8600_DU205NC16D02_8600   
1      6800_6800_0001_5_EPR91502_0      CONT_AWD_0001_6800_EPR91502_6800   
2     6800_6800_0001_20_EPR51502_0      CONT_AWD_0001_6800_EPR51502_6800   
3      6800_6800_0001_0_EPS21602_0      CONT_AWD_0001_6800_EPS21602_6800   
4     6800_6800_0001_31_EPS51301_0      CONT_AWD_0001_6800_EPS51301_6800   

   award_id_piid  modification_number  transaction_number  \
0              1                    0                   0   
1              1                    5                   0   
2              1                   20                   0   
3              1                    0                   0   
4              1                   31                   0   

   parent_award_agency_id                      parent_award_agency_name  \
0                    8600  HOUSING AND URBAN DEVELOPMENT, DEPARTMENT OF   
1                    6800               ENVIRONMENTAL PROTECTION AGENCY   
2                    6800               ENVIRONMENTAL PROTECTION AGENCY   
3                    6800               ENVIRONMENTAL PROTECTION AGENCY   
4                    6800               ENVIRONMENTAL PROTECTION AGENCY   

  parent_award_id_piid  parent_award_modification_number  \
0         DU205NC16D02                                 0   
1             EPR91502                                 0   
2             EPR51502                                 0   
3             EPS21602                                 0   
4             EPS51301                                 0   

   federal_action_obligation  total_dollars_obligated  \
0                  799750.00                546621.23   
1                       0.00                      NaN   
2                  750000.00               3042661.55   
3                 1149194.00                      NaN   
4                   99339.58                      NaN   

   base_and_exercised_options_value  current_total_value_of_award  \
0                         799750.00                     546621.23   
1                              0.00                           NaN   
2                         750000.00                   16600000.00   
3                        1389501.48                           NaN   
4                          99339.58                           NaN   

   base_and_all_options_value  potential_total_value_of_award  \
0                   799750.00                       546621.23   
1                        0.00                             NaN   
2                   750000.00                     16600000.00   
3                  1389501.48                             NaN   
4                    99339.58                             NaN   

   disaster_emergency_fund_codes_for_overall_award  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   outlayed_amount_funded_by_COVID-19_supplementals_for_overall_aw  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

   obligated_amount_funded_by_COVID-19_supplementals_for_overall_a  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  action_date  action_date_fiscal_year period_of_pe

### Columns of interest:

* contract_transaction_unique_key
* contract_award_unique_key
* awarding_agency_code
* awarding_sub_agency_code
* awarding_office_code
* funding_agency_code
* funding_sub_agency_code
* funding_office_code
* recipient_duns
* recipient_name
* recipient_doing_business_as_name
* recipient_parent_duns


## Step 2: Load data

Using dask with only columns of interest included.

In [3]:
columns = ['contract_transaction_unique_key', 'contract_award_unique_key', 'awarding_agency_code', 'awarding_sub_agency_code',\
           'awarding_office_code', 'funding_agency_code', 'funding_sub_agency_code', 'funding_office_code', 'recipient_duns', \
           'recipient_name', 'recipient_doing_business_as_name', 'recipient_parent_duns']

data = dd.read_csv('*.csv', usecols=columns, dtype={'awarding_sub_agency_code': 'object', 'funding_sub_agency_code': 'object', \
                                                    'recipient_doing_business_as_name': 'object', 'recipient_parent_duns': 'object',\
                                                    'funding_agency_code': 'object', 'awarding_office_code': 'object', \
                                                    'funding_office_code': 'object', 'recipient_duns': 'object'})

data.head()

contract_transaction_unique_key             contract_award_unique_key  \
0  8600_8600_0001_0_DU205NC16D02_0  CONT_AWD_0001_8600_DU205NC16D02_8600   
1      6800_6800_0001_5_EPR91502_0      CONT_AWD_0001_6800_EPR91502_6800   
2     6800_6800_0001_20_EPR51502_0      CONT_AWD_0001_6800_EPR51502_6800   
3      6800_6800_0001_0_EPS21602_0      CONT_AWD_0001_6800_EPS21602_6800   
4     6800_6800_0001_31_EPS51301_0      CONT_AWD_0001_6800_EPS51301_6800   

   awarding_agency_code awarding_sub_agency_code awarding_office_code  \
0                    86                     8600               86545A   
1                    68                     6800               68HE09   
2                    68                     6800               68HE05   
3                    68                     6800               68HE02   
4                    68                     6800               68HE05   

  funding_agency_code funding_sub_agency_code funding_office_code  \
0                 086                    8600              8630AX   
1                 068                    6800              68Y000   
2                 068                    6800              68U000   
3                 068                    6800              68R000   
4                 068                    6800              68U000   

  recipient_duns                                     recipient_name  \
0      831854596    MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1      112621987                                       INNOVATE INC   
2      079493355  RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
3      080265258                                       CAT-ENSOFTEK   
4      178599221                                TETRA TECH EM, INC.   

  recipient_doing_business_as_name recipient_parent_duns  
0                              NaN             831854596  
1                              NaN             112621987  
2                              NaN             079493355  
3                              NaN             080265258  
4                              NaN             045224250

### Step 2a: Save as parquet format

(So it's easier to reload if needed)

In [4]:
data.to_parquet('government_spend_dataset.parquet', engine='pyarrow')

(None,)

## Step 3: Get number of initial records

In [5]:
a = data.index.size.compute()

print("There are ", a, " initial records in this dataset.")

There are  4820290  initial records in this dataset.


## Step 4: Create list of unique suppliers (exact)

(and get number of unique suppliers)

In [6]:
suppliers_unique = data.recipient_name.unique().compute().reset_index()

print("There are ", len(suppliers_unique), " unique suppliers (exact match only).")

There are  153462  unique suppliers (exact match only).


In [7]:
suppliers_unique.head()

index                                     recipient_name
0      0    MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC
1      1                                       INNOVATE INC
2      2  RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...
3      3                                       CAT-ENSOFTEK
4      4                                TETRA TECH EM, INC.

### Step 4a: Does recipient_doing_business_as_name have any non-null values?

Answer: yes, hold off on this until next step

In [8]:
data.recipient_doing_business_as_name.dropna().compute()

23                MIT LINCOLN LABORATORY
29                                 U T C
39        SOFTWARE ENGINEERING INSTITUTE
52                    PACIFIC SCIENTIFIC
86                    PACIFIC SCIENTIFIC
                      ...               
18863    TOUCHSTONE NEURORECOVERY CENTER
18864                     PROVIDE A RIDE
18875             PETNET PHARMACEUTICALS
18882        MUSC DEPARTMENT OF MEDICINE
18895                 BIMBO BAKERIES USA
Name: recipient_doing_business_as_name, Length: 94562, dtype: object

## Step 5: Normalize text

Using:

 - list of punctuation from string package
 - custom regex to remove spaces between single character words
 - spacy English tokenizer
 - Snowball stemmer from NLTK
 - English stopwords from NLTK (updated to not use stopwords list)

Returning normalized tokens as list in additional column

In [10]:
# stop = set(stopwords.words('english'))
punct = string.punctuation

def remove_punct(text):
    text_str = str(text)
    for c in text_str:
        if c in punct:
            text_str = text_str.replace(c, '')
    return text_str

def remove_whitespace(text_str):
    text_str = re.sub(r'(?<=\s[A-Z0-9])[ ](?=[A-Z0-9]\s)', '', text_str)
    text_str = re.sub(r'(?<=^[A-Z0-9])[ ](?=[A-Z0-9])', '', text_str)
    text_str = re.sub(r'(?<=[A-Z0-9])[ ](?=[A-Z0-9]$)', '', text_str)
    return text_str

def normalize_text(text_str):
    wordlist = []
    for w in tokenizer(text_str):
        if not w.is_punct:
            wordlist.append(stemmer.stem(str(w).lower()))
    return wordlist

suppliers_unique['recipient_name_norm'] = suppliers_unique.recipient_name.apply(remove_punct).apply(remove_whitespace).apply(normalize_text)

suppliers_unique

index                                     recipient_name  \
0            0    MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1            1                                       INNOVATE INC   
2            2  RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
3            3                                       CAT-ENSOFTEK   
4            4                                TETRA TECH EM, INC.   
...        ...                                                ...   
153457  153457           AZORMED - GESTAO AMBIENTAL ACOREANA, LDA   
153458  153458                                      JEAN ROB & CO   
153459  153459                               WOODS SERVICES, INC.   
153460  153460                               FLOWERS & ASSOCIATES   
153461  153461                                       GARY BLUEMKE   

                                      recipient_name_norm  
0            [mortgag, assist, and, acquisit, group, llc]  
1                                            [innov, inc]  
2       [ryba, marin, construct, coterra, contract, se...  
3                                           [catensoftek]  
4                                  [tetra, tech, em, inc]  
...                                                   ...  
153457         [azorm,  , gestao, ambient, acoreana, lda]  
153458                                 [jean, rob,  , co]  
153459                                [wood, servic, inc]  
153460                                [flower,  , associ]  
153461                                     [gari, bluemk]  

[153462 rows x 3 columns]

In [11]:
suppliers_unique['recipient_name_str'] = suppliers_unique.recipient_name_norm.agg(lambda x: ','.join(map(str, x)))

suppliers_unique.recipient_name_str.value_counts()[suppliers_unique.recipient_name_str.value_counts() > 1].\
sort_values(ascending = False).to_frame(name = 'Additional matches')

Additional matches
innov,technolog,inc                        5
analyt,servic,inc                          5
macro,intern,inc                           4
shredit,usa,inc                            4
epsilon,system,solut,inc                   4
...                                      ...
angarai,intern,inc                         2
total,resourc,manag,inc                    2
hawk,enterpris,llc                         2
sbiw,inc                                   2
special,tactic,solut,inc                   2

[3734 rows x 1 columns]

There were 3,734 additional matches discovered through this simple text normalization.

In [14]:
suppliers_unique.recipient_name_str.unique().shape

(149545,)

With normalized text, **149,545** unique suppliers are found. These new normalized names can be joined to the original df.

In [16]:
data_norm = data.merge(suppliers_unique, on = 'recipient_name').compute()

data_norm

contract_transaction_unique_key  \
0                  8600_8600_0001_0_DU205NC16D02_0   
1                      6800_6800_0001_5_EPR91502_0   
2                      6800_6800_0002_3_EPR91502_0   
3                     6800_6800_0001_20_EPR51502_0   
4                      6800_6800_0001_0_EPS21602_0   
...                                            ...   
18914  7008_-NONE-_HSCG3216PSMA006_P00001_-NONE-_0   
18915    9700_-NONE-_SPE4A716D5005_0_-NONE-_-NONE-   
18916     3600_-NONE-_VA25815P0001_P00002_-NONE-_0   
18917    9700_-NONE-_SPE4A716D0005_0_-NONE-_-NONE-   
18918        1605_-NONE-_DOLJ089626807_31_-NONE-_0   

                         contract_award_unique_key  awarding_agency_code  \
0             CONT_AWD_0001_8600_DU205NC16D02_8600                    86   
1                 CONT_AWD_0001_6800_EPR91502_6800                    68   
2                 CONT_AWD_0002_6800_EPR91502_6800                    68   
3                 CONT_AWD_0001_6800_EPR51502_6800                    68   
4                 CONT_AWD_0001_6800_EPS21602_6800                    68   
...                                            ...                   ...   
18914  CONT_AWD_HSCG3216PSMA006_7008_-NONE-_-NONE-                    70   
18915                  CONT_IDV_SPE4A716D5005_9700                    97   
18916     CONT_AWD_VA25815P0001_3600_-NONE-_-NONE-                    36   
18917                  CONT_IDV_SPE4A716D0005_9700                    97   
18918    CONT_AWD_DOLJ089626807_1605_-NONE-_-NONE-                  1601   

      awarding_sub_agency_code awarding_office_code funding_agency_code  \
0                         8600               86545A                 086   
1                         6800               68HE09                 068   
2                         6800               68HE09                 068   
3                         6800               68HE05                 068   
4                         6800               68HE02                 068   
...                        ...                  ...                 ...   
18914                     7008                00032                 070   
18915                     97AS               SPE4A7                 097   
18916                     3600                00258                 036   
18917                     97AS               SPE4A7                 097   
18918                     1630                ETATP                1601   

      funding_sub_agency_code funding_office_code recipient_duns  \
0                        8600              8630AX      831854596   
1                        6800              68Y000      112621987   
2                        6800              68Y000      112621987   
3                        6800              68U000      079493355   
4                        6800              68R000      080265258   
...                       ...                 ...            ...   
18914                    7008               00032      785365693   
18915                    97AS              SPE4A7      089107726   
18916                    3600               00678      968237433   
18917                    97AS              SPE4A7      121360627   
18918                    1631               ETATP      145057522   

                                          recipient_name  \
0        MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1                                           INNOVATE INC   
2                                           INNOVATE INC   
3      RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
4                                           CAT-ENSOFTEK   
...                                                  ...   
18914                                   FLO-TECH, L.L.C.   
18915                  PRECI-MANUFACTURING, INCORPORATED   
18916                                       GARY BLUEMKE   
18917                           JANEL'S INDUSTRIES, INC.   
18918           UBC NATIONAL JOB CORPS TRAINING FUND INC   

      recipient_do

### Step 5a: Normalizing and matching recipient_doing_business_as_name column

Names also appear for some rows in recipient_doing_business_as_name. For rows where this appears, this can also be normalized and matched to normed values in recipient_name column.

In [24]:
suppliers_unique_rev = data_norm.recipient_doing_business_as_name.dropna().drop_duplicates().reset_index()

suppliers_unique_rev

index       recipient_doing_business_as_name
0       109                 MIT LINCOLN LABORATORY
1       136                                 BOEING
2       158                                  U T C
3       224         SOFTWARE ENGINEERING INSTITUTE
4       238        INTERNATIONAL BUSINESS MACHINES
...     ...                                    ...
8893  18571        GOLDEN LIVINGCENTER - BRIARWOOD
8894  18756                   AVALON PLACE WHARTON
8895  18758       BRIARWOOD NURSING&REHABILITATION
8896  18849                FAIRFIELD INN -MARRIOTT
8897  18875  HOPE NETWORK TRAUMATIC BRAIN INJURIES

[8898 rows x 2 columns]

In [26]:
suppliers_unique_rev['recipient_doing_business_norm'] = suppliers_unique_rev.recipient_doing_business_as_name.apply(remove_punct).\
apply(remove_whitespace).\
apply(normalize_text)

suppliers_unique_rev['recipient_doing_business_str'] = suppliers_unique_rev.recipient_doing_business_norm.agg(lambda x: ','.join(map(str, x)))

suppliers_unique_rev

index       recipient_doing_business_as_name  \
0       109                 MIT LINCOLN LABORATORY   
1       136                                 BOEING   
2       158                                  U T C   
3       224         SOFTWARE ENGINEERING INSTITUTE   
4       238        INTERNATIONAL BUSINESS MACHINES   
...     ...                                    ...   
8893  18571        GOLDEN LIVINGCENTER - BRIARWOOD   
8894  18756                   AVALON PLACE WHARTON   
8895  18758       BRIARWOOD NURSING&REHABILITATION   
8896  18849                FAIRFIELD INN -MARRIOTT   
8897  18875  HOPE NETWORK TRAUMATIC BRAIN INJURIES   

                recipient_doing_business_norm  \
0                  [mit, lincoln, laboratori]   
1                                       [boe]   
2                                       [utc]   
3                  [softwar, engin, institut]   
4                      [intern, busi, machin]   
...                                       ...   
8893       [golden, livingcent,  , briarwood]   
8894                 [avalon, place, wharton]   
8895            [briarwood, nursingrehabilit]   
8896               [fairfield, inn, marriott]   
8897  [hope, network, traumat, brain, injuri]   

           recipient_doing_business_str  
0                mit,lincoln,laboratori  
1                                   boe  
2                                   utc  
3                softwar,engin,institut  
4                    intern,busi,machin  
...                                 ...  
8893      golden,livingcent, ,briarwood  
8894               avalon,place,wharton  
8895         briarwood,nursingrehabilit  
8896             fairfield,inn,marriott  
8897  hope,network,traumat,brain,injuri  

[8898 rows x 4 columns]

In [33]:
data_norm_rev = data_norm.merge(suppliers_unique_rev, how = 'outer', on = ['recipient_doing_business_as_name'])

data_norm_rev

contract_transaction_unique_key  \
0                      8600_8600_0001_0_DU205NC16D02_0   
1                          6800_6800_0001_5_EPR91502_0   
2                          6800_6800_0002_3_EPR91502_0   
3                         6800_6800_0001_20_EPR51502_0   
4                          6800_6800_0001_0_EPS21602_0   
...                                                ...   
4820285  3600_-NONE-_VA24715A0007_P00001_-NONE-_-NONE-   
4820286       3600_-NONE-_VA25615G0083_0_-NONE-_-NONE-   
4820287       3600_-NONE-_VA25615A0093_0_-NONE-_-NONE-   
4820288      9700_-NONE-_N6830616A0073_0_-NONE-_-NONE-   
4820289  3600_-NONE-_VA79115D0009_P00001_-NONE-_-NONE-   

                    contract_award_unique_key  awarding_agency_code  \
0        CONT_AWD_0001_8600_DU205NC16D02_8600                    86   
1            CONT_AWD_0001_6800_EPR91502_6800                    68   
2            CONT_AWD_0002_6800_EPR91502_6800                    68   
3            CONT_AWD_0001_6800_EPR51502_6800                    68   
4            CONT_AWD_0001_6800_EPS21602_6800                    68   
...                                       ...                   ...   
4820285            CONT_IDV_VA24715A0007_3600                    36   
4820286            CONT_IDV_VA25615G0083_3600                    36   
4820287            CONT_IDV_VA25615A0093_3600                    36   
4820288           CONT_IDV_N6830616A0073_9700                    97   
4820289            CONT_IDV_VA79115D0009_3600                    36   

        awarding_sub_agency_code awarding_office_code funding_agency_code  \
0                           8600               86545A                 086   
1                           6800               68HE09                 068   
2                           6800               68HE09                 068   
3                           6800               68HE05                 068   
4                           6800               68HE02                 068   
...                          ...                  ...                 ...   
4820285                     3600                00247                 036   
4820286                     3600                00256                 036   
4820287                     3600                00256                 036   
4820288                     1700               N68306                 097   
4820289                     3600                00791                 036   

        funding_sub_agency_code funding_office_code recipient_duns  \
0                          8600              8630AX      831854596   
1                          6800              68Y000      112621987   
2                          6800              68Y000      112621987   
3                          6800              68U000      079493355   
4                          6800              68R000      080265258   
...                         ...                 ...            ...   
4820285                    3600               00508      622596158   
4820286                    3600               00580      079478766   
4820287                    3600               00580      795449045   
4820288                    1700              N68306      808502129   
4820289                    3600               00741      625294871   

                                            recipient_name  \
0          MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1                                             INNOVATE INC   
2                                             INNOVATE INC   
3        RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
4                                             CAT-ENSOFTEK   
...                                                    ...   
4820285                         GGNSC TUCKER BRIARWOOD LLC   
4820286                        SOUTHWEST LTC-WHARTON, LTD.   
4820287                  PINNACLE HEALTH FACILITIES XV, LP   
4820288                             OCEAN PROPERTIES, LTD.   
4820289                  

In [39]:
data_norm_rev['recipient_name_str'] = np.where(data_norm_rev.recipient_doing_business_str.str in suppliers_unique.recipient_name_str, \
                                                 data_norm_rev.recipient_doing_business_str, data_norm_rev.recipient_name_str)

data_norm_rev

## Step 6: Evaluating accuracy

### Step 6a: Alternative strategies

Instead of going the route of text normalization, conversion of sentence embeddings could be implemented to allow more flexibility and fine-tuning in the fuzzy matching procedure. With every word transformed into a numeric value, the resulting list per row could be compared to other rows in terms of a threshold cosine similarity or Levenshtein distance and fuzzy matching could be performed that way. The other benefit to this tactic would the ability for the analyst to control the matching threshold value. I initially thought about pursuing this strategy instead of or in addition to the normalization pipeline I ultimately used, but I did not have enough time to think about how to build a function that would do this pairwise calculation quickly between every possible pair of supplier names in this dataset, perhaps through k-means or other naive clustering analysis (resulting category labels could be appended to original df). It would be interesting to see how the accuracy of this strategy would compare to the strategy I ultimately used (the text normalization strategy may lead to an overcount of unique suppliers, while a strategy such as this, if the hyperparameters are not well tuned, may lead to an undercount of unique suppliers).

## Step 7: Newly answerable questions